src/storage/src/hummock/compactor/mod.rs

```diff
@@ -515,14 +518,55 @@ pub fn start_compactor(
-                                    let ((compact_task, table_stats, object_timestamps), _memory_tracker) =
-                                        compactor_runner::compact(
-                                            context.clone(),
-                                            compact_task,
-                                            rx,
-                                            Box::new(sstable_object_id_manager.clone()),
-                                            filter_key_extractor_manager.clone(),
-                                        ).await;
+                                    let existing_table_ids: HashSet<u32> = HashSet::from_iter(compact_task.existing_table_ids.clone());
+                                    let compact_table_ids = Vec::from_iter(
+                                        compact_task
+                                            .input_ssts
+                                            .iter()
+                                            .flat_map(|level| level.table_infos.iter())
+                                            .flat_map(|sst| sst.table_ids.clone())
+                                            .filter(|table_id| existing_table_ids.contains(table_id))
+                                            .sorted()
+                                            .unique(),
+                                    );
+
+                                     let ((compact_task, table_stats, object_timestamps), _memory_tracker) = match compaction_catalog_manager_ref.acquire(compact_table_ids.clone()).await {
+                                        Ok(compaction_catalog_agent_ref) => {
+                                            let acquire_table_ids: HashSet<StateTableId> = compaction_catalog_agent_ref.table_ids().collect();
+                                            if acquire_table_ids.len() != compact_table_ids.len() {
+                                                let diff = compact_table_ids.into_iter().collect::<HashSet<_>>()
+                                                    .symmetric_difference(&acquire_table_ids)
+                                                    .cloned()
+                                                    .collect::<Vec<_>>();
+                                                tracing::warn!(
+                                                    dif= ?diff,
+                                                    "Some table ids are not acquired."
+                                                );
+                                                let task_status = TaskStatus::ExecuteFailed;
+                                                (
+                                                    compact_done(compact_task, context.clone(), vec![], task_status),
+                                                    None,
+                                                )
+                                            } else {
```

Comment:
Let me share my opinion on this.
`CompactionCatalogManager` is designed to manage the catalog, and build the CompactionCatalogAgent from the catalog. (Using only `SchemaFilterKeyExtractor`), I think, for the `compact` interface, it makes more sense to pass the CompactionCatalogAgent for the `compact` interface, I think it makes more sense, it's better for testing, and we can disguise different types of `FilterKeyExtractor`.

In response to your question, can I move the legitimacy checking of table_ids inside the compact function so that we don't have to set the state of the compact task externally explicitly?

Reply:
Revert the externally checking and separate the `compact` to two function @hzxa21 PTAL